### DISCUSSION TITLE
Multistage docker build with venv error: "ModuleNotFoundError: No module named 'flask'"

### DISCUSSION THREAD

**Discussion by rosario-vandebron** [tiangolo/meinheld-gunicorn-docker#141] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched in the documentation/README.
- [X] I already searched in Google "How to do X" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs/README and didn't find an answer.

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```Text
FROM tiangolo/meinheld-gunicorn:python3.9-alpine3.13-2024-03-11 AS app
ENV PATH="/venv/bin:$PATH"
ENV PYTHONPATH="/venv/bin:$PATH"

ENV PORT 5000

EXPOSE 5000

ENV MODULE_NAME="bmprelive.apps.internal.manage"
ENV GUNICORN_CONF="/app/bmprelive/apps/internal/gunicorn_conf.py"

# copy in Python environment
COPY --from=builder /venv /venv
```

### Description

I am trying to build a multistage docker image starting from `tiangolo/meinheld-gunicorn:python3.9-alpine3.13-2024-03-11` . I've installed all my dependencies using `pip` and `virtualenv`. After all dependencies are installed I simply do in my dockerfile

```
FROM tiangolo/meinheld-gunicorn:python3.9-alpine3.13-2024-03-11 AS app
ENV PATH="/venv/bin:$PATH"
ENV PYTHONPATH="/venv/bin:$PATH"

ENV PORT 5000

EXPOSE 5000

ENV MODULE_NAME="bmprelive.apps.internal.manage"
ENV GUNICORN_CONF="/app/bmprelive/apps/internal/gunicorn_conf.py"

# copy in Python environment
COPY --from=builder /venv /venv
```

but then when I try to run the container I see this error:

```
File "/app/bmprelive/apps/internal/__init__.py", line 2, in <module>
    from flask import Flask
ModuleNotFoundError: No module named 'flask'
[2024-11-18 21:40:40 +0000] [9] [INFO] Worker exiting (pid: 9)
[2024-11-18 21:40:40 +0000] [10] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
    worker.init_process()
  File "/usr/local/lib/python3.9/site-packages/gunicorn/workers/base.py", line 134, in init_process
    self.load_wsgi()
```
Looking at the interpreter used I've noticed that it is using the system python and not the venv one. Further investigation confirmed that gunicorn has been installed globally from your image. 

Now the question is: Is there a way to use multistage build along with your image? Is there a way to tell which python interpreter the workers should use ? Or maybe I can do a multistage build without venv but I have no experience with it. Can you help?

### Operating System

Linux

### Operating System Details

Docker

### Python Version

3.9

**Comment by tiangolo** [comment, marked as answer]
This image is not designed to do multi stage builds, the intention was to simplify as much as possible the setup you had to do, but if you're already doing multi stage builds, you would probably benefit more from doing it all from scratch on your side, including installing the server. 🤓
